File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/ExchangeRateAdapter.java
package com.blogspot.e_kanivets.moneytracker.adapter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import com.blogspot.e_kanivets.moneytracker.MtApp;
import com.blogspot.e_kanivets.moneytracker.controller.FormatController;
import com.blogspot.e_kanivets.moneytracker.databinding.ViewExchangeRateBinding;
import com.blogspot.e_kanivets.moneytracker.entity.ExchangeRatePair;

import java.util.List;

import javax.inject.Inject;

public class ExchangeRateAdapter extends BaseAdapter {
    @Inject
    FormatController formatController;

    private Context context;
    private List<ExchangeRatePair> exchangeRates;

    public ExchangeRateAdapter(Context context, List<ExchangeRatePair> exchangeRates) {
        this.context = context;
        this.exchangeRates = exchangeRates;
        MtApp.get().getAppComponent().inject(ExchangeRateAdapter.this);
    }

    @Override
    public int getCount() {
        return exchangeRates.size();
    }

    @Override
    public ExchangeRatePair getItem(int position) {
        return exchangeRates.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @SuppressLint("SetTextI18n")
    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {
        ViewExchangeRateBinding binding;

        if (convertView == null) {
            LayoutInflater layoutInflater = LayoutInflater.from(context);

            binding = ViewExchangeRateBinding.inflate(layoutInflater, parent, false);
            convertView = binding.getRoot();

            convertView.setTag(binding);
        } else {
            binding = (ViewExchangeRateBinding) convertView.getTag();
        }

        ExchangeRatePair rate = getItem(position);

        binding.tvFromCurrency.setText(rate.getFromCurrency());
        binding.tvToCurrency.setText(rate.getToCurrency());
        binding.tvAmountBuy.setText(formatController.formatPrecisionNone(rate.getAmountBuy()));
        binding.tvAmountSell.setText(formatController.formatPrecisionNone(rate.getAmountSell()));

        return convertView;
    }
}

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/ExchangeRateAdapter.kt
package com.blogspot.e_kanivets.moneytracker.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import com.blogspot.e_kanivets.moneytracker.MtApp
import com.blogspot.e_kanivets.moneytracker.controller.FormatController
import com.blogspot.e_kanivets.moneytracker.databinding.ViewExchangeRateBinding
import com.blogspot.e_kanivets.moneytracker.entity.ExchangeRatePair
import javax.inject.Inject

class ExchangeRateAdapter(
    private val context: Context,
    private val exchangeRates: List<ExchangeRatePair>
) : BaseAdapter() {

    @Inject lateinit var formatController: FormatController

    init {
        MtApp.instance.appComponent.inject(this)
    }

    override fun getCount(): Int = exchangeRates.size

    override fun getItem(position: Int): ExchangeRatePair = exchangeRates[position]

    override fun getItemId(position: Int): Long = position.toLong()

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val binding: ViewExchangeRateBinding
        var view = convertView
        if (view == null) {
            val layoutInflater = LayoutInflater.from(context)
            binding = ViewExchangeRateBinding.inflate(layoutInflater, parent, false)
            view = binding.root
            view.tag = binding
        } else {
            binding = view.tag as ViewExchangeRateBinding
        }
        val rate = getItem(position)
        binding.tvFromCurrency.text = rate.fromCurrency
        binding.tvToCurrency.text = rate.toCurrency
        binding.tvAmountBuy.text = formatController.formatPrecisionNone(rate.amountBuy)
        binding.tvAmountSell.text = formatController.formatPrecisionNone(rate.amountSell)
        return view
    }
}
